alternator/streams.cc

```diff
@@ -589,44 +679,40 @@ struct shard_iterator {
     static auto constexpr inclusive_marker = 'I';
     static auto constexpr exclusive_marker = 'i';
 
-    shard_id shard;
+    static auto constexpr uuid_str_length = 36;
+
+    utils::UUID table;
     utils::UUID threshold;
+    shard_id shard;
     bool inclusive;
 
-    shard_iterator(shard_id i, utils::UUID th, bool inclusive)
-        : shard(i)
-        , threshold(th)
-        , inclusive(inclusive)
-    {}
-    shard_iterator(utils::UUID arn, db_clock::time_point ts, cdc::stream_id i, utils::UUID th, bool inclusive)
-        : shard(arn, ts, i)
+    shard_iterator(const sstring&);
+
+    shard_iterator(utils::UUID t, shard_id i, utils::UUID th, bool inclusive)
+        : table(t)
         , threshold(th)
+        , shard(i)
         , inclusive(inclusive)
     {}
-    shard_iterator(const sstring&);
-
     friend std::ostream& operator<<(std::ostream& os, const shard_iterator& id) {
         boost::io::ios_flags_saver fs(os);
-        return os << (id.inclusive ? inclusive_marker : exclusive_marker) << std::hex << id.threshold
+        return os << (id.inclusive ? inclusive_marker : exclusive_marker) << std::hex
+            << id.table
+            << ':' << id.threshold
             << ':' << id.shard
             ;
     }
-private:
-    shard_iterator(const sstring&, size_t);
 };
 
-shard_iterator::shard_iterator(const sstring& s) 
-    : shard_iterator(s, s.find(':'))
-{}
-
-shard_iterator::shard_iterator(const sstring& s, size_t i)
-    : shard(s.substr(i + 1))
-    , threshold(s.substr(1, i))
+shard_iterator::shard_iterator(const sstring& s)
+    : table(s.substr(1, uuid_str_length))
+    , threshold(s.substr(2 + uuid_str_length, uuid_str_length))
+    , shard(s.substr(3 + 2 * uuid_str_length))
+    , inclusive(s[0] == inclusive_marker)
```

Comment:
If s comes from the user, you can't do s[0], it might not exist. This is why the old code used s.at(0) before thinking of using s[0]. You can use s.at(0) here, and s[0] in the constructor function body, after you already know s.at(0) didn't throw.

Reply:
We already did substr(1, ...), which will throw if the string is smaller that 2, so the [] there is already checked. But pro format, sure.

Reply:
You're right, I forgot that substr() throws.